### PR TITLE
mds: reverse MDSMap encoding of max_xattr_size/bal_rank_mask

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
In 5 hours: https://pulpito.ceph.com/?branch=wip-batrick-testing-20240215.160715

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2024-02-16_01:25:08-fs:upgrade:mds_upgrade_sequence-wip-batrick-testing-20240215.160715-distro-default-smithi/

File: qa/cephfs/overrides/ignorelist_health.yaml [hunk 1]
```diff
@@ -8,6 +8,7 @@ overrides:
       - MDS_DAMAGE
       - MDS_DEGRADED
       - MDS_FAILED
+      - MDS_INSUFFICIENT_STANDBY
```
Comment (vshankar) [inline qa/cephfs/overrides/ignorelist_health.yaml:11]: This has started to show up for some reason (debugging now) in pacific too - https://pulpito.ceph.com/lflores-2024-02-15_17:32:20-rados-wip-yuri10-testing-2024-02-08-0854-pacific-distro-default-smithi/7561657/

There other host of warnings are PG_DEGRADED which are new too in fs:upgrade too.

Comment (vshankar) [reply qa/cephfs/overrides/ignorelist_health.yaml:11]: Volume got created at:

```
2024-02-15T18:22:52.001 DEBUG:teuthology.orchestra.run.smithi067:> sudo /home/ubuntu/cephtest/cephadm --image docker.io/ceph/ceph:v16.2.4 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 343b5320-cc2e-11ee-95ba-87774f69a715 -- bash -c 'ceph fs volume create cephfs --placement=4'
```

Warning report check from mon.0:

```
2024-02-15T18:22:55.693+0000 7f2c37c18700  7 mon.smithi067@0(leader).mds e1 print_map
e2
enable_multiple, ever_enabled_multiple: 1,1
compat: compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,8=no anchor table,9=file layout v2,10=snapreal
m v2}
legacy client fscid: 1

Filesystem 'cephfs' (1)
fs_name cephfs
epoch   2
flags   12
created 2024-02-15T18:22:55.695260+0000
modified        2024-02-15T18:22:55.695260+0000
tableserver     0
root    0
session_timeout 60
session_autoclose       300
max_file_size   1099511627776
required_client_features        {}
last_failure    0
last_failure_osd_epoch  0
compat  compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,8=no anchor table,9=file layout v2,10=snapreal
m v2}
max_mds 1
in
up      {}
failed
damaged
stopped
data_pools      [3]
metadata_pool   2
inline_data     disabled
balancer
standby_count_wanted    0



2024-02-15T18:22:55.694+0000 7f2c37c18700  4 mon.smithi067@0(leader).mds e1 prepare_command done, r=0
2024-02-15T18:22:55.694+0000 7f2c37c18700 10 mon.smithi067@0(leader).paxosservice(mdsmap 1..1) dispatch forced immediate propose
2024-02-15T18:22:55.694+0000 7f2c37c18700 10 mon.smithi067@0(leader).paxosservice(mdsmap 1..1) propose_pending
2024-02-15T18:22:55.694+0000 7f2c37c18700 10 mon.smithi067@0(leader).mds e1 encode_pending e2
2024-02-15T18:22:55.694+0000 7f2c37c18700 10 mon.smithi067@0(leader) e2 log_health updated 2 previous 0
2024-02-15T18:22:55.694+0000 7f2c37c18700 -1 log_channel(cluster) log [ERR] : Health check failed: 1 filesystem is offline (MDS_ALL_DOWN)
```

~3s of delay when the MDS was probably created., but this is expected as you mention, but what else changed to have this warning consistently show up now...

Comment (vshankar) [reply qa/cephfs/overrides/ignorelist_health.yaml:11]: The MDS were deployed ~9s later.

```
2024-02-15T18:23:02.315+0000 7f47dc273780  1 -- 172.21.15.174:0/3228562887 >> [v2:172.21.15.174:3300/0,v1:172.21.15.174:6789/0] conn(0x561b3a2e8c00 msgr2=0x561b3a330300 secure :-1 s=STATE_CONNECTION_ESTABLISHED l=1).mark_down
2024-02-15T18:23:02.315+0000 7f47dc273780  1 --2- 172.21.15.174:0/3228562887 >> [v2:172.21.15.174:3300/0,v1:172.21.15.174:6789/0] conn(0x561b3a2e8c00 0x561b3a330300 secure :-1 s=READY pgs=32 cs=0 l=1 rev1=1 rx=0x561b3afc0750 tx=0x561b3afca120).stop
2024-02-15T18:23:02.315+0000 7f47dc273780  1 -- 172.21.15.174:0/3228562887 shutdown_connections
2024-02-15T18:23:02.315+0000 7f47dc273780  1 --2- 172.21.15.174:0/3228562887 >> [v2:172.21.15.67:3300/0,v1:172.21.15.67:6789/0] conn(0x561b3a2e9000 0x561b3a330800 unknown :-1 s=CLOSED pgs=0 cs=0 l=0 rev1=0 rx=0 tx=0).stop
2024-02-15T18:23:02.315+0000 7f47dc273780  1 --2- 172.21.15.174:0/3228562887 >> [v2:172.21.15.174:3300/0,v1:172.21.15.174:6789/0] conn(0x561b3a2e8c00 0x561b3a330300 unknown :-1 s=CLOSED pgs=32 cs=0 l=1 rev1=1 rx=0 tx=0).stop
2024-02-15T18:23:02.315+0000 7f47dc273780  1 -- 172.21.15.174:0/3228562887 >> 172.21.15.174:0/3228562887 conn(0x561b3a2e8800 msgr2=0x561b3a3ce000 unknown :-1 s=STATE_NONE l=0).mark_down
2024-02-15T18:23:02.315+0000 7f47dc273780  1 -- 172.21.15.174:0/3228562887 shutdown_connections
2024-02-15T18:23:02.315+0000 7f47dc273780  1 -- 172.21.15.174:0/3228562887 wait complete.
2024-02-15T18:23:02.315+0000 7f47dc273780  0 set uid:gid to 167:167 (ceph:ceph)
2024-02-15T18:23:02.315+0000 7f47dc273780  0 ceph version 16.2.4 (3cbe25cde3cfa028984618ad32de9edc4c1eaed0) pacific (stable), process ceph-mds, pid 7
```

cephadm got a _bit_ slower?

Comment (vshankar) [reply qa/cephfs/overrides/ignorelist_health.yaml:11]: There is also

> Replacing daemon mds.cephfs.smithi038.jgrryv as rank 0

warning in the fs:upgrade test which is also hitting in pacific: https://github.com/ceph/ceph/pull/55523#issuecomment-1947304053

Comment (vshankar) [reply qa/cephfs/overrides/ignorelist_health.yaml:11]: Probably a fallout from https://github.com/ceph/ceph/pull/54312/

**Comment by lxbsz** [comment]
The kclient fix https://patchwork.kernel.org/project/ceph-devel/list/?series=827346.

**Review by vshankar — APPROVED** [review]

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#20th-Feb-2024

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#20th-Feb-2024

**Pull request merged**